tests/test_router_quote_in_given_out.py

```diff
@@ -0,0 +1,243 @@
+
+        transmuter_token_pair = transmuter_token_data[1]
+
+        denom_in = transmuter_token_pair[1]
+        denom_out = transmuter_token_pair[0]
+
+        # This is the max error tolerance of 5% that we allow.
+        # Arbitrarily hand-picked to avoid flakiness.
+        error_tolerance = 0.05
+
+        # Get denom in precision.
+        denom_out_precision = conftest.get_denom_exponent(denom_out)
+
+        # Get denom out data to retrieve precision and price 
+        denom_in_data = conftest.shared_test_state.chain_denom_to_data_map.get(denom_in)
+        denom_in_precision = denom_in_data.get("exponent")
+        
+        # Compute spot price scaling factor.
+        spot_price_scaling_factor = Decimal(10)**denom_out_precision / Decimal(10)**denom_in_precision
+
+        # Compute expected spot prices
+        out_base_in_quote_price = Decimal(denom_in_data.get("price"))
+        expected_out_base_in_quote_price = 1 / out_base_in_quote_price
+        
+        # Compute expected token out
+        expected_token_in = int(amount) * expected_out_base_in_quote_price
+
+        # Run the quote test
+        quote = self.run_quote_test(environment_url, amount + denom_out, denom_in, EXPECTED_LATENCY_UPPER_BOUND_MS)
+
+        # Validate transmuter was in route
+        assert Quote.is_transmuter_in_single_route(quote.route) is True
+
+        # Validate the quote test
+        self.validate_quote_test(quote, amount, denom_out, spot_price_scaling_factor, expected_out_base_in_quote_price, expected_token_in, error_tolerance)
+    def run_quote_test(self, environment_url, token_out, token_in, expected_latency_upper_bound_ms, expected_status_code=200) -> QuoteExactAmountOutResponse:
+        """
+        Runs exact amount out test for the /router/quote endpoint with the given input parameters.
+
+        Does basic validation around response status code and latency
+
+        Returns quote for additional validation if needed by client
+
+        Validates:
+        - Response status code is as given or default 200
+        - Latency is under the given bound
+        """
+
+        sqs_service = conftest.SERVICE_MAP[environment_url]
+
+        start_time = time.time()
+        response = sqs_service.get_exact_amount_out_quote(token_out, token_in)
+        elapsed_time_ms = (time.time() - start_time) * 1000
+
+        assert response.status_code == expected_status_code, f"Error: {response.text}"
+        assert expected_latency_upper_bound_ms > elapsed_time_ms, f"Error: latency {elapsed_time_ms} exceeded {expected_latency_upper_bound_ms} ms, token in {token_in} and token out {token_out}" 
+
+        response_json = response.json()
+
+        # Return route for more detailed validation
+        return QuoteExactAmountOutResponse(**response_json)
+
+    def validate_quote_test(self, quote, expected_amount_out_str, expected_denom_out, spot_price_scaling_factor, expected_out_base_in_quote_price, expected_token_in, error_tolerance):
+        """
```

Comment:
I know that this did not exist for "out given in" but, given new functionality, I think it would be helpful to validate the route format.

That is, that by following the returned route from the API, we go through the pools that indeed contain the specified tokens and end up correctly in the token in.

Additionally, since route pool may now contain either "token_in_denom" or "token_out_denom", this validation would be helpful.

I'm happy if you would like to add it in a separate PR.